pkg/reconciler/configmappropagation/configmappropagation.go

```diff
@@ -0,0 +1,314 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package configmappropagation
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"knative.dev/eventing/pkg/apis/configs/v1alpha1"
+	configslisters "knative.dev/eventing/pkg/client/listers/configs/v1alpha1"
+	"knative.dev/eventing/pkg/logging"
+	"knative.dev/eventing/pkg/reconciler"
+	"knative.dev/eventing/pkg/reconciler/configmappropagation/resources"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/tracker"
+)
+
+const (
+	// Name of the corev1.Events emitted from the reconciliation process
+	configMapPropagationReconcileError                  = "configMapPropagationReconcileError"
+	configMapPropagationUpdateStatusFailed              = "ConfigMapPropagationStatusFailed"
+	configMapPropagationReadinessChanged                = "ConfigMapPropagationReadinessChanged"
+	configMapPropagationPropagateSingleConfigMapFailed  = "ConfigMapPropagationPropagateSingleConfigMapFailed"
+	configMapPropagationPropagateSingleConfigMapSucceed = "ConfigMapPropagationPropagateSingleConfigMapSucceed"
+)
+
+type Reconciler struct {
+	*reconciler.Base
+
+	// Listers index properties about resources
+	configMapPropagationLister configslisters.ConfigMapPropagationLister
+	configMapLister            corev1listers.ConfigMapLister
+
+	tracker tracker.Interface
+}
+
+var configMapGVK = corev1.SchemeGroupVersion.WithKind("ConfigMap")
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the ConfigMapPropagation resource
+// with the current status of the resource.
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logging.FromContext(ctx).Error("invalid resource key")
+		return nil
+	}
+
+	// Get the ConfigMapPropagation resource with this namespace/name
+	original, err := r.configMapPropagationLister.ConfigMapPropagations(namespace).Get(name)
+
+	if apierrs.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logging.FromContext(ctx).Info("ConfigMapPropagation key in work queue no longer exists")
+		return nil
+	} else if err != nil {
```

Comment:
(nit) a switch may be more idiomatic here
```golang
switch {
case apierrs.IsNotFound(err):
    // ...
case err != nil:
    // ...
}
```

Reply:
For this part, I followed the same way as other reconcilers did (like pkg/reconciler/broker/broker.go, pkg/reconciler/trigger/trigger.go).

Reply:
Is that really a reason not to write idiomatic code where possible ?

Reply:
I am just wondering if that is more idiomatic, why no other places do that. Do you have some readability review guidelines for this?

Reply:
Not at all, it may be completely arbitrary. I'm just used to seeing apierrors handled that way in the Kubernetes codebase, and I'm probably alergic to `else if`s :) Feel free to discard.

Reply:
@antoineco There are multiple idioms. This one came from the original kubernetes sample controller years ago (which in the meantime has [changed a bit](https://github.com/kubernetes/sample-controller/blob/bd4ff3d8adf21a05e468f522a14fabc7c4e70f65/controller.go#L252)).

If you're interested in using switches here, that's fair, but we should use switches everywhere this code is duplicated. Create an issue in eventing and serving explaining why you want to change it and create PRs.

Another option is to talk to @n3wscott who IIUC is working on generating this code. You could get your preferred idiom into the generator template :smile: